email/blocker.go

```diff
@@ -0,0 +1,215 @@
+		}
+
+		// loop all emails and parse them
+		for _, email := range toBlock {
+			err = func() (err error) {
+				lock := abuseDB.NewLock(email.UID)
+				err = lock.Lock()
+				if err != nil {
+					return errors.AddContext(err, "could not acquire lock")
+				}
+				defer func() {
+					unlockErr := lock.Unlock()
+					if unlockErr != nil {
+						err = errors.Compose(err, errors.AddContext(unlockErr, "could not release lock"))
+						return
+					}
+				}()
+
+				// parse the email
+				result, err := b.blockReport(email.ParseResult)
+				if err != nil {
+					return errors.AddContext(err, "could not block report")
+				}
+
+				// sanity check we have a result for every skylink
+				if len(result) != len(email.ParseResult.Skylinks) {
+					return errors.New("block result not defined for every skylink")
+				}
+
+				// update the email
+				err = abuseDB.UpdateNoLock(email,
+					bson.D{
+						{"$set", bson.D{
+							{"blocked", true},
+							{"blockResult", result},
+						}},
+					},
+				)
+				if err != nil {
+					return errors.AddContext(err, "could not update email")
+				}
+				return nil
+			}()
+			if err != nil {
+				logger.Errorf("Failed to parse email %v, error %v", email.UID, err)
+			}
+		}
+	}
+}
+
+// blockReport will block all skylinks from the given abuse report.
+func (b *Blocker) blockReport(report database.AbuseReport) ([]string, error) {
+	var results []string
+	for _, skylink := range report.Skylinks {
+		result := func() string {
+			// build the request
+			req, err := b.buildBlockRequest(skylink, report)
+			if err != nil {
+				return fmt.Sprintf("failed to build request, err: %v", err.Error())
+			}
+
+			// execute the request
+			b.staticLogger.Debugf("blocking %v", skylink)
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				return fmt.Sprintf("failed to execute request, err: %v", err.Error())
+			}
+
+			// handle the response
+			switch resp.StatusCode {
+			case http.StatusOK, http.StatusNoContent:
+				return database.AbuseStatusBlocked
+			default:
+				respBody, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					return fmt.Sprintf("failed to read response body, err: %v", err.Error())
+				}
+				resp.Body.Close()
+				return fmt.Sprintf("failed to block skylink, status %v response: %v", resp.Status, string(respBody))
+			}
+		}()
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+// buildBlockRequest buils a request to be sent to the blocker API using the
+// provided input.
+func (b *Blocker) buildBlockRequest(skylink string, report database.AbuseReport) (*http.Request, error) {
+	// build the request body
+	reqBody := BlockPOST{
+		Skylink:  skylink,
+		Reporter: report.Reporter,
+		Tags:     report.Tags,
+	}
+
+	// build the request
+	reqBodyBytes, err := json.Marshal(reqBody)
+	if err != nil {
+		return nil, err
+	}
+	reqBodyBuffer := bytes.NewBuffer(reqBodyBytes)
+
+	url := fmt.Sprintf("http://%s/block", b.staticBlockerApiUrl)
```

Comment:
Here we directly call the URL specified by `BLOCKER_API_URL` but we also add `http://` to it. This is a bit strange. It means the person writing the config needs to write the URL in a weird format.

Let's change this to either be the full URL or just the IP and port. `BLOCKER_IP` and `BLOCKER_PORT` are already used in malware-scanner but if you prefer a full URL I can change it there as well.

Reply:
Ok fair, I like the api url though, just simplifies things. I dropped the `http://`

Reply:
Changed my mind, reused `BLOCKER_IP` and `BLOCKER_PORT` to limit the amount of duplicated env vars but I think `BLOCKER_API_URL` makes more sense so maybe we should clean that up after.

Reply:
Sure, I don't mind using `BLOCKER_API_URL`, as long as we do the same everywhere (e.g. malware-scanner).